x-pack/plugins/observability_solution/apm_data_access/server/services/get_host_names/index.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { estypes } from '@elastic/elasticsearch';
+import { rangeQuery } from '@kbn/observability-plugin/server';
+import { RegisterParams } from '../register_services';
+
+export interface ServicesHostNamesParams {
+  query: estypes.QueryDslQueryContainer;
+  from: number;
+  to: number;
+  limit: number;
+}
+
+export function createGetHostNames(params: RegisterParams) {
+  return async ({ from: timeFrom, to: timeTo, query, limit }: ServicesHostNamesParams) => {
+    const { apmIndices, esClient } = await params.getResourcesForServices();
```

Comment:
I think we should stay away from patterns like this, and expect the consumer to pass in asynchronously resolved dependencies, such as getting the APM indices. If we don't do that, we'll end up with a ton of small requests - I see this happen with e.g. the ML plugin that does a ton of small requests because it encapsulates privilege and other checks in each service call. For instance: https://github.com/elastic/kibana/issues/161229

Reply:
Yeah, I was just having second thoughts about this approach.

Reply:
though it's kind of weird asking the consumer to pass the APM indices to the `apm_data_acess` plugin.

Reply:
it is a little weird, and there might be a better way (e.g. to have a per-request service that does per-request caching, but that also might create some risks).

Reply:
I can't really see the risks of getting the APM indices when calling a service in the `apm_data_access`,  what could happen? if we just get the APM indices there, do you think it would still end up in the same situation from ML plugin?

I'm more concerned about the fact I'm using the internal user to both getting apm indices and typed esClient.

Reply:
One downside is potentially requesting APM indices twice.

Reply:
> One downside is potentially requesting APM indices twice.

This is my (only) concern. It might end up being more than twice however. Especially in the context of e.g. rule executors this becomes more important. My experience is that if we abstract away async operations, people lose awareness of what is happening as part of the service call, and it leads to performance bottlenecks. I'd be less concerned about this if we have really good insight in production bottlenecks, but we don't.

Reply:
For the lack of better alternatives, I'll leave it up to the consumer the send the APM indices for now. It's bad DX but for the time being it's better than end up in situation where it might lead to performance problems.